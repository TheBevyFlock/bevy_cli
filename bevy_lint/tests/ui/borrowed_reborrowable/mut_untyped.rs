//! This tests the `borrowed_reborrowable` lint, specifically when triggered on the `MutUntyped` type.

#![feature(register_tool)]
#![register_tool(bevy)]
#![deny(bevy::borrowed_reborrowable)]

use bevy::ecs::change_detection::MutUntyped;
use bevy::ecs::ptr::PtrMut;
use bevy::prelude::*;

// OK: Lint does not apply to immutable references
fn immutable_reference(_param: &MutUntyped) {
    // ...
}

//~| HELP: use `MutUntyped` instead
//~v ERROR: parameter takes `&mut MutUntyped` instead of a re-borrowed `MutUntyped`
fn mutable_reference(_param: &mut MutUntyped) {
    // ...
}

//~| HELP: use `MutUntyped` instead
//~v ERROR: parameter takes `&mut MutUntyped` instead of a re-borrowed `MutUntyped`
fn mutable_reference_return<'a>(_param: &'a mut MutUntyped) -> usize {
    123
}

// OK: Lint does not apply when return type relies on reference lifetime
fn mutable_reference_bounded_return<'a>(param: &'a mut MutUntyped) -> PtrMut<'a> {
    param.as_mut()
}

// OK: Lint does not apply when return type relies on reference lifetime
fn mutable_reference_bounded_return_complex<'a>(
    param: &'a mut MutUntyped,
) -> Vec<(usize, PtrMut<'a>)> {
    vec![(1, param.as_mut())]
}

fn main() {
    fn some_system(world: &mut World) {
        let mut param: MutUntyped = world
            .spawn(Name::new("test"))
            .into_mut::<Name>()
            .unwrap()
            .into();
        immutable_reference(&param);
        mutable_reference(&mut param);
        _ = mutable_reference_return(&mut param);
        _ = mutable_reference_bounded_return(&mut param);
        _ = mutable_reference_bounded_return_complex(&mut param);
    }

    App::new().add_systems(Update, some_system).run();
}
