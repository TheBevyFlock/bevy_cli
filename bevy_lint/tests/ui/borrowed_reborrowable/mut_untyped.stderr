error: parameter takes `&mut MutUntyped` instead of a re-borrowed `MutUntyped`
  --> tests/ui/borrowed_reborrowable/mut_untyped.rs:18:22
   |
18 | fn mutable_reference(_param: &mut MutUntyped) {
   |                      ^^^^^^^^^^^^^^^^^^^^^^^ help: use `MutUntyped` instead: `mut _param: bevy::bevy_ecs::change_detection::MutUntyped<'_>`
   |
note: the lint level is defined here
  --> tests/ui/borrowed_reborrowable/mut_untyped.rs:5:9
   |
5  | #![deny(bevy::borrowed_reborrowable)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: parameter takes `&mut MutUntyped` instead of a re-borrowed `MutUntyped`
  --> tests/ui/borrowed_reborrowable/mut_untyped.rs:24:33
   |
24 | fn mutable_reference_return<'a>(_param: &'a mut MutUntyped) -> usize {
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use `MutUntyped` instead: `mut _param: bevy::bevy_ecs::change_detection::MutUntyped<'_>`

error: aborting due to 2 previous errors

